File: learn/graph/disjoint-set-union/disjoint_set_union.go


type DisjointSetUnion struct {
	parent []int
	size   []int
}

func NewDisjointSetUnion(n int) *DisjointSetUnion {
	dsu := &DisjointSetUnion{
		parent: make([]int, n),
		size:   make([]int, n),
	}
	for i := 0; i < n; i++ {
		dsu.parent[i] = i
		dsu.size[i] = 1
	}
	return dsu
}

func (dsu *DisjointSetUnion) Find(v int) int {
	if v == dsu.parent[v] {
		return v
	}
	dsu.parent[v] = dsu.Find(dsu.parent[v])
	return dsu.parent[v]
}

func (dsu *DisjointSetUnion) Union(a, b int) {
	a = dsu.Find(a)
	b = dsu.Find(b)

	if a != b {
		if dsu.size[a] < dsu.size[b] {
			a, b = b, a
		}
		dsu.parent[b] = a
		dsu.size[a] += dsu.size[b]
	}
}

File: learn/graph/disjoint-set-union/disjoint_set_union.rs
pub struct DisjointSetUnion {
    parent: Vec<usize>,
    size: Vec<usize>,
}

impl DisjointSetUnion {
    pub fn new(n: usize) -> Self {
        let parent: Vec<usize> = (0..n).collect();
        let size = vec![1; n];
        DisjointSetUnion { parent, size }
    }

    pub fn find(&mut self, v: usize) -> usize {
        if v == self.parent[v] {
            return v;
        }
        self.parent[v] = self.find(self.parent[v]);
        self.parent[v]
    }

    pub fn union(&mut self, mut a: usize, mut b: usize) {
        a = self.find(a);
        b = self.find(b);

        if a != b {
            if self.size[a] < self.size[b] {
                std::mem::swap(&mut a, &mut b);
            }
            self.parent[b] = a;
            self.size[a] += self.size[b];
        }
    }
}
